## Pandas DataFrame I

In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
os.chdir('F:\DS tut\datasets')

In [3]:
cars_data = pd.read_csv('Toyota.csv')

In [4]:
#removing extra index value column
cars_data = pd.read_csv('Toyota.csv', index_col=0)

In [5]:
#removing junk values
cars_data = pd.read_csv('Toyota.csv', na_values=['??', '????'])

In [6]:
#shallow copy - refers to original object, changes in original will also be reflected in the copy
samp = cars_data.copy(deep=False)

In [7]:
samp

Unnamed: 0  Price   Age       KM FuelType     HP  MetColor  Automatic  \
0              0  13500  23.0  46986.0   Diesel   90.0       1.0          0   
1              1  13750  23.0  72937.0   Diesel   90.0       1.0          0   
2              2  13950  24.0  41711.0   Diesel   90.0       NaN          0   
3              3  14950  26.0  48000.0   Diesel   90.0       0.0          0   
4              4  13750  30.0  38500.0   Diesel   90.0       0.0          0   
...          ...    ...   ...      ...      ...    ...       ...        ...   
1431        1431   7500   NaN  20544.0   Petrol   86.0       1.0          0   
1432        1432  10845  72.0      NaN   Petrol   86.0       0.0          0   
1433        1433   8500   NaN  17016.0   Petrol   86.0       0.0          0   
1434        1434   7250  70.0      NaN      NaN   86.0       1.0          0   
1435        1435   6950  76.0      1.0   Petrol  110.0       0.0          0   

        CC  Doors  Weight  
0     2000  three    1165  
1     2000      3    1165  
2     2000      3    1165  
3     2000      3    1165  
4     2000      3    1170  
...    ...    ...     ...  
1431  1300      3    1025  
1432  1300      3    1015  
1433  1300      3    1015  
1434  1300      3    1015  
1435  1600      5    1114  

[1436 rows x 11 columns]

In [8]:
#deep copy - no reference to original object
cars_data1 = cars_data.copy(deep=True)

In [9]:
#removing extra index value column
cars_data1 = pd.read_csv('Toyota.csv', index_col=0)
#removing junk values
cars_data1 = pd.read_csv('Toyota.csv', na_values=['??', '????'])

In [10]:
#getting indices range
cars_data1.index

RangeIndex(start=0, stop=1436, step=1)

In [11]:
#getting attributes (column names)
cars_data1.columns

Index(['Unnamed: 0', 'Price', 'Age', 'KM', 'FuelType', 'HP', 'MetColor',
       'Automatic', 'CC', 'Doors', 'Weight'],
      dtype='object')

In [12]:
#size
cars_data1.size

15796

In [13]:
#shape - rows, columns
cars_data1.shape

(1436, 11)

In [14]:
#memory usage of each column
cars_data1.memory_usage()

Index           128
Unnamed: 0    11488
Price         11488
Age           11488
KM            11488
FuelType      11488
HP            11488
MetColor      11488
Automatic     11488
CC            11488
Doors         11488
Weight        11488
dtype: int64

In [15]:
#no. of axes/array dimensions
cars_data1.ndim

2

#### Indexing and slicing

In [16]:
#head() - by default returns first 5 rows
cars_data1.head()

Unnamed: 0  Price   Age       KM FuelType    HP  MetColor  Automatic    CC  \
0           0  13500  23.0  46986.0   Diesel  90.0       1.0          0  2000   
1           1  13750  23.0  72937.0   Diesel  90.0       1.0          0  2000   
2           2  13950  24.0  41711.0   Diesel  90.0       NaN          0  2000   
3           3  14950  26.0  48000.0   Diesel  90.0       0.0          0  2000   
4           4  13750  30.0  38500.0   Diesel  90.0       0.0          0  2000   

   Doors  Weight  
0  three    1165  
1      3    1165  
2      3    1165  
3      3    1165  
4      3    1170

In [17]:
cars_data1.head(8)

Unnamed: 0  Price   Age       KM FuelType    HP  MetColor  Automatic    CC  \
0           0  13500  23.0  46986.0   Diesel  90.0       1.0          0  2000   
1           1  13750  23.0  72937.0   Diesel  90.0       1.0          0  2000   
2           2  13950  24.0  41711.0   Diesel  90.0       NaN          0  2000   
3           3  14950  26.0  48000.0   Diesel  90.0       0.0          0  2000   
4           4  13750  30.0  38500.0   Diesel  90.0       0.0          0  2000   
5           5  12950  32.0  61000.0   Diesel  90.0       0.0          0  2000   
6           6  16900  27.0      NaN   Diesel   NaN       NaN          0  2000   
7           7  18600  30.0  75889.0      NaN  90.0       1.0          0  2000   

   Doors  Weight  
0  three    1165  
1      3    1165  
2      3    1165  
3      3    1165  
4      3    1170  
5      3    1170  
6      3    1245  
7      3    1245

In [18]:
#tail - by default returns last 5 rows
cars_data1.tail()

Unnamed: 0  Price   Age       KM FuelType     HP  MetColor  Automatic  \
1431        1431   7500   NaN  20544.0   Petrol   86.0       1.0          0   
1432        1432  10845  72.0      NaN   Petrol   86.0       0.0          0   
1433        1433   8500   NaN  17016.0   Petrol   86.0       0.0          0   
1434        1434   7250  70.0      NaN      NaN   86.0       1.0          0   
1435        1435   6950  76.0      1.0   Petrol  110.0       0.0          0   

        CC Doors  Weight  
1431  1300     3    1025  
1432  1300     3    1015  
1433  1300     3    1015  
1434  1300     3    1015  
1435  1600     5    1114

In [19]:
cars_data1.tail(8)

Unnamed: 0  Price   Age       KM FuelType     HP  MetColor  Automatic  \
1428        1428   8450  72.0      NaN   Petrol   86.0       NaN          0   
1429        1429   8950  78.0  24000.0   Petrol   86.0       1.0          1   
1430        1430   8450  80.0  23000.0   Petrol   86.0       0.0          0   
1431        1431   7500   NaN  20544.0   Petrol   86.0       1.0          0   
1432        1432  10845  72.0      NaN   Petrol   86.0       0.0          0   
1433        1433   8500   NaN  17016.0   Petrol   86.0       0.0          0   
1434        1434   7250  70.0      NaN      NaN   86.0       1.0          0   
1435        1435   6950  76.0      1.0   Petrol  110.0       0.0          0   

        CC Doors  Weight  
1428  1300     3    1015  
1429  1300     5    1065  
1430  1300     3    1015  
1431  1300     3    1025  
1432  1300     3    1015  
1433  1300     3    1015  
1434  1300     3    1015  
1435  1600     5    1114

In [20]:
#scalar value look ups
#at: label based lookup - [row no., attribute name]
#returns value of the attribute name at the goven row no.
cars_data1.at[5, 'KM']

61000.0

In [21]:
#iat: integer based lookups - [row no., col no.]
#returns value of attribute at the given row no. 
cars_data1.iat[2, 5]

90.0

In [22]:
#loc: access group of rows or columns
#returns rows with the given attribute
cars_data1.loc[:, 'FuelType']
#will return all the rows of attribute 'fuel type'

0       Diesel
1       Diesel
2       Diesel
3       Diesel
4       Diesel
         ...  
1431    Petrol
1432    Petrol
1433    Petrol
1434       NaN
1435    Petrol
Name: FuelType, Length: 1436, dtype: object

## Pandas DataFrames II

In [23]:
cars_data1.loc[5:64, ['HP', 'KM', 'Price']]

HP        KM  Price
5    90.0   61000.0  12950
6     NaN       NaN  16900
7    90.0   75889.0  18600
8   192.0   19700.0  21500
9     NaN   71138.0  12950
10  192.0   31461.0  20950
11  192.0   43610.0  19950
12  192.0   32189.0  19600
13  192.0   23000.0  21500
14  192.0   34131.0  22500
15    NaN   18739.0  22000
16  192.0   34000.0  22750
17  110.0   21716.0  17950
18  110.0   25563.0  16750
19  110.0   64359.0  16950
20  110.0   67660.0  15950
21  110.0   43905.0  16950
22  110.0   56349.0  15950
23  110.0   32220.0  16950
24  110.0   25813.0  16250
25  110.0   28450.0  15950
26  110.0   34545.0  17495
27  110.0   41415.0  15750
28  110.0   44142.0  16950
29  110.0   11090.0  17950
30   97.0    9750.0  12950
31   97.0   35199.0  15750
32   97.0   29510.0  15950
33   97.0   32692.0  14950
34   97.0   41000.0  15500
35   97.0   43000.0  15750
36   97.0   25000.0  15950
37   97.0   10000.0  14950
38   97.0   25329.0  15750
39   97.0   27500.0  14750
40   97.0   49059.0  13950
41   97.0   44068.0  16750
42   97.0   46961.0  13950
43   90.0  110404.0  16950
44   90.0  100250.0  16950
45   90.0   84000.0  19000
46   90.0   79375.0  17950
47   97.0   75048.0  15800
48   90.0   72215.0  17950
49  192.0   64982.0  21950
50   90.0   62636.0  17950
51   97.0   57086.0  15750
52  110.0   56000.0  20500
53  192.0   49866.0  21950
54  110.0   49163.0  15500
55  110.0   45725.0  13250
56   97.0   43210.0  15250
57   97.0   43000.0  15250
58  110.0   39704.0  18950
59  110.0   38950.0  15999
60   97.0   37400.0  14950
61  110.0   37177.0  16500
62  110.0   36544.0  18750
63  110.0   33511.0  17950
64   97.0       NaN  17950

In [24]:
#checking datatypes
cars_data1.dtypes

Unnamed: 0      int64
Price           int64
Age           float64
KM            float64
FuelType       object
HP            float64
MetColor      float64
Automatic       int64
CC              int64
Doors          object
Weight          int64
dtype: object

In [25]:
#count of unique data types
#DataFrame.get_dtype_counts() is deprecated since pandas version 0.25.0
#use DataFrame.dtypes.value_counts() instead
cars_data1.dtypes.value_counts()

int64      5
float64    4
object     2
dtype: int64

In [26]:
#selecting data based on data types
cars_data1.select_dtypes(exclude=[object])

Unnamed: 0  Price   Age       KM     HP  MetColor  Automatic    CC  \
0              0  13500  23.0  46986.0   90.0       1.0          0  2000   
1              1  13750  23.0  72937.0   90.0       1.0          0  2000   
2              2  13950  24.0  41711.0   90.0       NaN          0  2000   
3              3  14950  26.0  48000.0   90.0       0.0          0  2000   
4              4  13750  30.0  38500.0   90.0       0.0          0  2000   
...          ...    ...   ...      ...    ...       ...        ...   ...   
1431        1431   7500   NaN  20544.0   86.0       1.0          0  1300   
1432        1432  10845  72.0      NaN   86.0       0.0          0  1300   
1433        1433   8500   NaN  17016.0   86.0       0.0          0  1300   
1434        1434   7250  70.0      NaN   86.0       1.0          0  1300   
1435        1435   6950  76.0      1.0  110.0       0.0          0  1600   

      Weight  
0       1165  
1       1165  
2       1165  
3       1165  
4       1170  
...      ...  
1431    1025  
1432    1015  
1433    1015  
1434    1015  
1435    1114  

[1436 rows x 9 columns]

In [27]:
#info(): returns concise summary of a dataframe
'''
- data type of index
- data type of columns
- count of non-null values
- memory usage
'''
cars_data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1436 non-null   int64  
 1   Price       1436 non-null   int64  
 2   Age         1336 non-null   float64
 3   KM          1421 non-null   float64
 4   FuelType    1336 non-null   object 
 5   HP          1430 non-null   float64
 6   MetColor    1286 non-null   float64
 7   Automatic   1436 non-null   int64  
 8   CC          1436 non-null   int64  
 9   Doors       1436 non-null   object 
 10  Weight      1436 non-null   int64  
dtypes: float64(4), int64(5), object(2)
memory usage: 123.5+ KB


In [28]:
#unique(): used to find unique elemnts of a column
print(np.unique(cars_data1['KM']))

[  1.  15. 225. ...  nan  nan  nan]


In [29]:
print(np.unique(cars_data1['HP']))

[ 69.  71.  72.  73.  86.  90.  97.  98. 107. 110. 116. 192.  nan  nan
  nan  nan  nan  nan]


In [30]:
print(np.unique(cars_data1['Doors']))

['2' '3' '4' '5' 'five' 'four' 'three']


In [31]:
print(np.unique(cars_data1['Automatic']))

[0 1]


## Pandas Dataframes III

In [32]:
cars_data = pd.read_csv('Toyota.csv', na_values=["??", "????"])

In [33]:
cars_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1436 non-null   int64  
 1   Price       1436 non-null   int64  
 2   Age         1336 non-null   float64
 3   KM          1421 non-null   float64
 4   FuelType    1336 non-null   object 
 5   HP          1430 non-null   float64
 6   MetColor    1286 non-null   float64
 7   Automatic   1436 non-null   int64  
 8   CC          1436 non-null   int64  
 9   Doors       1436 non-null   object 
 10  Weight      1436 non-null   int64  
dtypes: float64(4), int64(5), object(2)
memory usage: 123.5+ KB


In [34]:
'''
converting variable's data type
astype(): convert from one dtype to another
'''
cars_data['MetColor'] = cars_data['MetColor'].astype('object')

In [35]:
cars_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1436 non-null   int64  
 1   Price       1436 non-null   int64  
 2   Age         1336 non-null   float64
 3   KM          1421 non-null   float64
 4   FuelType    1336 non-null   object 
 5   HP          1430 non-null   float64
 6   MetColor    1286 non-null   object 
 7   Automatic   1436 non-null   int64  
 8   CC          1436 non-null   int64  
 9   Doors       1436 non-null   object 
 10  Weight      1436 non-null   int64  
dtypes: float64(3), int64(5), object(3)
memory usage: 123.5+ KB


In [36]:
'''
nbytes(): used to get the total bytes consumed by the elements of the columns
'''
cars_data['FuelType'].nbytes

11488

In [37]:
cars_data['FuelType'].astype('category').nbytes

1460

In [38]:
cars_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1436 entries, 0 to 1435
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  1436 non-null   int64  
 1   Price       1436 non-null   int64  
 2   Age         1336 non-null   float64
 3   KM          1421 non-null   float64
 4   FuelType    1336 non-null   object 
 5   HP          1430 non-null   float64
 6   MetColor    1286 non-null   object 
 7   Automatic   1436 non-null   int64  
 8   CC          1436 non-null   int64  
 9   Doors       1436 non-null   object 
 10  Weight      1436 non-null   int64  
dtypes: float64(3), int64(5), object(3)
memory usage: 123.5+ KB


In [39]:
'''
cleaning column 'Doors'
replace(): replace a value with desired value
'''
cars_data['Doors'].replace('three', 3, inplace=True)
cars_data['Doors'].replace('four', 4, inplace=True)
cars_data['Doors'].replace('five', 5, inplace=True)

In [40]:
cars_data['Doors']=cars_data['Doors'].astype('int64')
print(np.unique(cars_data['Doors']))

[2 3 4 5]


In [41]:
'''
detecting missing values
isnull.sum(): count of missing values in each column
'''
cars_data.isnull().sum()

Unnamed: 0      0
Price           0
Age           100
KM             15
FuelType      100
HP              6
MetColor      150
Automatic       0
CC              0
Doors           0
Weight          0
dtype: int64

## Control structures

Create a new column called 'Price_Class':

```python
cars_data1.insert(10, "Price_Class", "")
```

We will classify the prices into 3 categories: 'Low'(<8450), 'High'(>11950) and 'Medium'(between 8450 and 11950).


Using if-elif-else:
``` python   
for i in range(0, len(cars_data1['Price']), 1):
    if (cars_data1['Price'][i]<=8450):
        cars_data1['Price_Class'][i]="Low"
    elif (cars_data1['Price'][i]>11950):
        cars_data1['Price_Class'][i]="High"
    else:
        cars_data1['Price_Class'][i]="Medium"
    
```
   

Using while loop:
```python
i=0

while i<len(cars_data1['Price']):
    if (cars_data1['Price'][i]<=8450):
        cars_data1['Price_Class'][i]="Low"
    elif (cars_data1['Price'][i]>11950):
        cars_data1['Price_Class'][i]="High"
    else:
        cars_data1['Price_Class'][i]="Medium"
    i = i+1
```

Series.value_counts(): retunrs series containing count of unique values.
```python
cars_data1['Price_Class'].value_counts()
```


## Functions

### Function with a single argument

Create a new column called 'Age_Converted'

```python
cars_data1.insert(11, 'Age_Converted',0)
```

Define a function called 'c_convert':
```python
def c_convert(val):
    val_converted = val/12
    return val_converted

cars_data1["Age_Converted"]=c_convert(cars_data1['Age'])
cars_data1["Age_Converted"]=round(cars_data1["Age_Converted"], 1)
```

### Function with multiple inputs and outputs

• Functions in Python takes multiple input objects but return only one object as output.

• However lists, tuples or dictionaries can be used to return multiple outputs as required.

• A multiple input multiple output function c_convert has been defined.

• The function takes in two inputs.

• The output is returned in the form of a list.

```python
def c_convert(val1, val2):
    val_converted = val1/12
    ratio = val2/val1
    return [val_converted, ratio]

cars_data1["Age_Converted"], cars_data1["KM_per_month"] = \c_convert(cars_data1["Age"], cars_data1["KM"])
```

• Here, Age and KM columns of the data set are input to the function.

• The outputs are assigned to ‘Age_Converted’ and ‘km_per_month’.
